config/src/config/network_config.rs

```diff
@@ -340,58 +340,95 @@ impl Default for RateLimitConfig {
     }
 }
 
-pub type TrustedPeerSet = HashMap<PeerId, TrustedPeer>;
+pub type PeerSet = HashMap<PeerId, Peer>;
+
+// TODO: Combine with RoleType?
+/// Represents the Role that a peer plays in the network ecosystem rather than the type of node.
+/// Determines how nodes are connected to other nodes, and how discovery views them.
+///
+/// Rules for upstream nodes via Peer Role:
+///
+/// Validator -> Always upstream if not Validator else P2P
+/// PreferredUpstream -> Always upstream, overriding any other discovery
+/// ValidatorFullNode -> Always upstream for incoming connections (including other ValidatorFullNodes)
+/// Upstream -> Upstream, if no ValidatorFullNode or PreferredUpstream.  Useful for initial seed discovery
+/// Unknown -> Undiscovered peer, likely due to a non-mutually authenticated connection always downstream
+#[derive(Clone, Copy, Debug, Deserialize, PartialEq, Serialize)]
+pub enum PeerRole {
+    Validator = 0,
+    PreferredUpstream,
+    ValidatorFullNode,
+    Upstream,
+    Unknown,
+}
+
+impl Default for PeerRole {
+    /// Default to least trusted
+    fn default() -> Self {
+        PeerRole::Unknown
+    }
+}
 
 /// Represents a single seed configuration for a seed peer
 #[derive(Clone, Debug, Default, Deserialize, PartialEq, Serialize)]
 #[serde(default)]
-pub struct TrustedPeer {
-    pub keys: HashSet<x25519::PublicKey>,
+pub struct Peer {
     pub addresses: Vec<NetworkAddress>,
+    pub keys: HashSet<x25519::PublicKey>,
+    pub role: PeerRole,
 }
 
-impl TrustedPeer {
+impl Peer {
     /// Combines `Vec<NetworkAddress>` keys with the `HashSet` given
     pub fn new(
         addresses: Vec<NetworkAddress>,
         mut keys: HashSet<x25519::PublicKey>,
-    ) -> TrustedPeer {
+        role: PeerRole,
+    ) -> Peer {
         let addr_keys = addresses
             .iter()
             .filter_map(NetworkAddress::find_noise_proto);
         keys.extend(addr_keys);
-        TrustedPeer { addresses, keys }
+        Peer {
+            addresses,
+            keys,
+            role,
+        }
     }
 
-    /// Combines two `TrustedPeer`.  Note: Does not merge duplicate addresses
-    pub fn extend(&mut self, other: TrustedPeer) {
+    /// Combines two `Peer`.  Note: Does not merge duplicate addresses
+    /// TODO: Instead of rejecting, maybe pick one of the roles?
+    pub fn extend(&mut self, other: Peer) -> Result<(), Error> {
+        crate::config::invariant(
+            self.role != other.role,
+            format!(
+                "Roles don't match self {:?} vs other {:?}",
+                self.role, other.role
+            ),
+        )?;
```

Comment:
maybe it would make sense to `max(self.role, other.role)`, like pick the more important role type when merging

Reply:
Maybe, I'm unsure this is really a migration function, because it's not preferred outside of the merging of `seed_addrs` and `seeds`, since it doesn't do a very good job of merging the addresses.